Mathlib/Tactic/Peel.lean

```diff
@@ -0,0 +1,215 @@
+/-
+Copyright (c) 2023 Jireh Loreaux. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+import Mathlib.Order.Filter.Basic
+import Mathlib.Tactic.Basic
+
+/-!
+# The `peel` tactic
+
+`peel h with idents*` tries to apply `forall_imp` (or `Exists.imp`, or `Filter.Eventually.mp`,
+`Filter.Frequently.mp` and `Filter.eventually_of_forall`) with the argument `h` and uses `idents*`
+to introduce variables with the supplied names.
+
+Alternatively, one can provide a numeric argument as in `peel 4 h` which will peel 4 quantifiers off
+the expressions automatically name the introduced variables.
+
+In addition, the user may supply a term `e` via `... using e` in order to close the goal
+immediately. In particular, `peel h using e` is equivalent to `peel h; exact e`. The `using` syntax
+may be paired with any of the other features of `peel`.
+-/
+
+open Lean Expr Meta Elab Tactic Mathlib.Tactic
+
+/--
+Peels matching quantifiers off of a given term and the goal and introduces the relevant variables.
+
+Given `p q : ℕ → Prop`, `h : ∀ x, p x`, and a goal `⊢ : ∀ x, q x`, the tactic `peel h with h' x`
+will introduce `x : ℕ`, `h' : p x` into the context and the new goal will be `⊢ q x`. This works
+with `∃`, as well as `∀ᶠ` and `∃ᶠ`, and it can even be applied to a sequence of quantifiers. Note
+that this is a logically weaker setup, so using this tactic is not always feasible.
+
+For a more complex example, given a hypothesis and a goal:
+```
+h : ∀ ε > (0 : ℝ), ∃ N : ℕ, ∀ n ≥ N, 1 / (n + 1 : ℝ) < ε
+⊢ ∀ ε > (0 : ℝ), ∃ N : ℕ, ∀ n ≥ N, 1 / (n + 1 : ℝ) ≤ ε
+```
+(which differ only in `<`/`≤`), applying `peel h with h_peel ε hε N n hn` will yield a tactic state:
+```
+h : ∀ ε > (0 : ℝ), ∃ N : ℕ, ∀ n ≥ N, 1 / (n + 1 : ℝ) < ε
+ε : ℝ
+hε : 0 < ε
+N n : ℕ
+hn : N ≤ n
+h_peel : 1 / (n + 1 : ℝ) < ε
+⊢ 1 / (n + 1 : ℝ) ≤ ε
+```
+and the goal can be closed with `exact h_peel.le`.
+Note that in this example, `h` and the goal are logically equivalent statements, but `peel`
+*cannot* be immediately applied to show that the goal implies `h`.
+
+`peel` can take an optional numeric argument prior to the supplied hypothesis; in which case it will
+autoname the introduced variables, but they may be inaccessible. However, in this case the user must
+still introduce a single name such as `with h_peel` for the new hypothesis.
+
+In addition, `peel` supports goals of the form `(∀ x, p x) ↔ ∀ x, q x`, or likewise for any
```

Comment:
Maybe `peel` should mention that the `congr!` tactic can also apply to such goals, but `peel` has the benefit that it only applies to quantifiers (so accounting and predictability are easier for example).

Reply:
good point. I added the `↔` support primarily for pedagogical purposes, but that's good documentation.